src/OperationProcessor.ts

```diff
@@ -75,580 +17,160 @@ export interface OperationProcessor {
-      }
-    }
+    return this.operationStore.delete(transactionNumber);
   }
 
   /**
    * Resolve the given DID to its DID Doducment.
    * @param did The DID to resolve. e.g. did:sidetree:abc123.
    * @returns DID Document of the given DID. Undefined if the DID is deleted or not found.
+   *
+   * Iterate over all operations in blockchain-time order extending the
+   * the operation chain while checking validity.
    */
   public async resolve (did: string): Promise<DidDocument | undefined> {
-    const didUniqueSuffix = did.substring(this.didMethodName.length);
-
-    await this.processDeferredOperationsOfDid(didUniqueSuffix);
+    let didDocument: DidDocument | undefined;
+    let previousOperation: Operation | undefined;
 
-    if (this.deletedDids.has(did)) {
-      return undefined;
-    }
+    const didUniqueSuffix = did.substring(this.didMethodName.length);
 
-    const latestVersion = await this.last(didUniqueSuffix);
+    const didOps = await this.operationStore.get(didUniqueSuffix);
+    for (const operation of didOps) {
+      if (await this.isValid(operation, previousOperation, didDocument)) {
+        didDocument = this.getUpdatedDocument(didDocument, operation);
+        previousOperation = operation;
 
-    // lastVersion === undefined implies we do not know about the did
-    if (latestVersion === undefined) {
-      return undefined;
+        if (operation.type === OperationType.Delete) {
+          break;
+        }
+      }
     }
 
-    return this.lookup(latestVersion);
+    return didDocument;
   }
 
   /**
-   * Returns the DID Document for a given version identifier.
+   * Is an operation valid in the context of a resolve() when extending
+   * operation chain for a DID. When this function is called, resolve() has
+   * constructed a valid operation that starts with some Create operation and
+   * ends with previousOperation, resulting the DID Document 'currentDidDocument'.
+   * This function checks if parameter operation can be used to legitimately
+   * extend this chain.
    */
-  public async lookup (versionId: VersionId): Promise<DidDocument | undefined> {
-    // Version id is also the operation hash that produces the document
-    const opHash = versionId;
-
-    const opInfo = this.opHashToInfo.get(opHash);
-
-    // We don't know anything about this operation
-    if (opInfo === undefined) {
-      return undefined;
-    }
+  private async isValid (operation: Operation, previousOperation: Operation | undefined, currentDidDocument: DidDocument | undefined):
+    Promise<boolean> {
 
-    // Construct the operation using a CAS lookup
-    const op = await this.operationStore.lookup(opHash);
+    if (operation.type === OperationType.Create) {
 
-    if (this.isInitialVersion(opInfo)) {
-      if (opInfo.status === OperationStatus.Invalid) {
-        return undefined;
+      // If either of these is defined, then we have seen a previous create operation.
+      if (previousOperation || currentDidDocument) {
```

Comment:
What if there are only two operations in chronological order: U1, C1, in this logic C1 is consider invalid, but I should be valid.

Reply:
U1, C1 should leave C1 in a valid state just as you expect. When U1 is being processed we check if previous is defined - in this case U1 would be invalid, so prevoiusOperation would be null, so this if condition would not be taken.

Reply:
You are right.